truffle/src/main/java/org/jruby/truffle/runtime/subsystems/ObjectSpaceManager.java

```diff
@@ -135,4 +142,38 @@ private static void runFinalizers(RubyContext context, FinalizerReference finali
         return handlers;
     }
 
+    public void traceAllocationsStart() {
+        tracingAssumptionActivations++;
+
+        if (tracingAssumptionActivations == 1) {
+            tracingAssumption.getAssumption().invalidate();
```

Comment:
Ah I see, this line only invalidates the current Assumption but does not replace it like CyclicAssumption.invalidate() does.